### ISSUE TITLE
TypeScript: can't infer prop names and types when contains any type prop

### ISSUE THREAD

**Issue by wonderful-panda** [vuejs/vue#6841]
### Version
2.5.2

### Reproduction link
[https://github.com/wonderful-panda/vue-sandbox](https://github.com/wonderful-panda/vue-sandbox)

### Steps to reproduce
Compile this code by tsc

```typescript
import Vue, { VNode } from "vue";

const NgComponent = Vue.extend({
    name: "NgComponent",
    props: {
        foo: { type: String },
        bar: {}    // any types are accepted
    },
    render(h): VNode {
        return h("div", this.foo);
    }
});
```

or clone from repro link and `npm install && npm run build`

### What is expected?
Successfully compiled.

And `this.foo` is treated as string, `this.bar` is treated as any.

### What is actually happening?
Failed to compile.

```
src/index.ts(10,30): error TS2339: Property 'foo' does not exist on type 'CombinedVueInstance<Vue, {}, {}, {}, Readonly<{}>>'.
```

---
Workaround:

```ts
import Vue, { VNode } from "vue";

Vue.extend({
    name: "Workaround",
    props: {
        foo: { type: String },
        bar: {} as (() => any)   // any types are accepted
    },
    render(h): VNode {
        return h("div", this.foo);
    }
});
```

**Comment by ktsn** [comment]
Indeed it does not work if it has any typed props definition. I'm not sure about the reason, though.

@HerringtonDarkholme @DanielRosenwasser
I would appreciate if you can look into it.

**Comment by DanielRosenwasser** [comment]
I'm currently not by a computer but I'll guess a bit. Right now it's probably an assignability error for props in general, but we can do better there in future versions of TypeScript. I'm currently on a branch where this works better.

The thing is that even on this branch, the best inference I can get is `{}`.

**Comment by HerringtonDarkholme** [comment]
With #6850 , you can do this.

```ts
import Vue, { VNode } from "vue";

Vue.extend({
    name: "Workaround",
    props: {
        foo: { type: String },
        bar: {} as Prop<any>
    },
    render(h): VNode {
        return h("div", this.foo);
    }
});
```

I think inferring `any` from empty prop type is too lenient. A lot of options will have empty field, which results in empty type parameter. If we make inference default to `any`, there will be too many errors hidden by unintended `any`.

**Comment by yyx990803** [comment]
Closing this one - please keep an eye on #6856 (to be merged for 2.6)
